File: master/option.cc
#include <iostream>
#include <cstring>

#include "defs.h"
#include "sos.h"
#include "option.h"
#include "utils.h"

bool option::initialized_ = false ;
struct option::optdesc option::optdesc_ [256] ;


#define	OPTION_INIT		do { \
				    if (! initialized_) \
					static_init () ; \
				} while (false)			// no ";"
#define	RESET			do { \
				    optcode_ = MO_None ; \
				    optlen_ = 0 ; \
				    optval_ = 0 ; \
				} while (false)			// no ";"
#define	ALLOC_COPY(f,m,l)	do { \
				    f = new byte [(l)] ; \
				    std::memcpy (f, (m), (l)) ; \
				} while (false)			// no ";"
#define	COPY_VAL(p)		do { \
				    byte *b ; \
				    if (optlen_ > (int) sizeof staticval_) \
					b = optval_ = new byte [optlen_] ; \
				    else \
				    { \
					optval_ = 0 ; \
					b = staticval_ ; \
				    } \
				    std::memcpy (b, (p), optlen_) ; \
				} while (false)			// no ";"
#define	CHECK_OPTCODE(c)	do { \
				    if ((c) < 0 || (c) >= NTAB (optdesc_) || optdesc_ [c].format == OF_NONE) \
					throw 20 ; \
				} while (false)			// no ";"
#define	CHECK_OPTLEN(c,l)	do { \
				    if ((l) < optdesc_ [(c)].minlen \
					    || (l) > optdesc_ [(c)].maxlen) \
					throw 20 ; \
				} while (false)			// no ";"


/******************************************************************************
 * Utilities
 */

struct stbin
{
    byte bin [sizeof (option::uint)] ;
    int len ;
} ;

stbin *uint_to_byte (option::uint val)
{
    static stbin stbin ;
    int shft ;

    // translate in network byte order, without leading null bytes
    stbin.len = 0 ;
    for (shft = sizeof val - 1 ; shft >= 0 ; shft--)
    {
        byte b ;

        b = (val >> (shft * 8)) & 0xff ;
        if (stbin.len != 0 || b != 0)
            stbin.bin [stbin.len++] = b ;
    }
    return &stbin ;
}

/******************************************************************************
 * Constructors and destructors
 */

// default constructor
option::option ()
{
    OPTION_INIT ;
    RESET ;
}

option::option (optcode_t optcode)
{
    OPTION_INIT ;
    CHECK_OPTCODE (optcode) ;
    RESET ;
    optcode_ = optcode ;
}

option::option (optcode_t optcode, void *optval, int optlen)
{
    OPTION_INIT ;
    CHECK_OPTCODE (optcode) ;
    CHECK_OPTLEN (optcode, optlen) ;
    RESET ;
    optcode_ = optcode ;
    optlen_ = optlen ;
    COPY_VAL (optval) ;
}

option::option (optcode_t optcode, option::uint optval)
{
    stbin *stbin ;

    stbin = uint_to_byte (optval) ;
    OPTION_INIT ;
    CHECK_OPTCODE (optcode) ;
    CHECK_OPTLEN (optcode, stbin->len) ;
    RESET ;
    optcode_ = optcode ;
    optlen_ = stbin->len ;
    COPY_VAL (stbin->bin) ;
}

// copy constructor
option::option (const option &o)
{
    OPTION_INIT ;
    *this = o ;
    if (optval_)
	COPY_VAL (o.optval_) ;
}

// copy assignment constructor
option &option::operator= (const option &o)
{
    OPTION_INIT ;
    if (this != &o)
    {
	*this = o ;
	if (optval_)
	    COPY_VAL (o.optval_) ;
    }
    return *this ;
}

// default destructor
option::~option ()
{
    if (optval_)
	delete optval_ ;
}


void option::static_init (void)
{
    initialized_ = true ;

    for (int i = 0 ; i < NTAB (optdesc_) ; i++)
	optdesc_ [i].format = OF_NONE ;

    optdesc_ [MO_Content_Format].format = OF_OPAQUE ;
    optdesc_ [MO_Content_Format].minlen = 0 ;
    optdesc_ [MO_Content_Format].maxlen = 8 ;

    optdesc_ [MO_Etag].format = OF_OPAQUE ;
    optdesc_ [MO_Etag].minlen = 1 ;
    optdesc_ [MO_Etag].maxlen = 8 ;

    optdesc_ [MO_Location_Path].format = OF_STRING ;
    optdesc_ [MO_Location_Path].minlen = 0 ;
    optdesc_ [MO_Location_Path].maxlen = 255 ;

    optdesc_ [MO_Location_Query].format = OF_STRING ;
    optdesc_ [MO_Location_Query].minlen = 0 ;
    optdesc_ [MO_Location_Query].maxlen = 255 ;

    optdesc_ [MO_Max_Age].format = OF_UINT ;
    optdesc_ [MO_Max_Age].minlen = 0 ;
    optdesc_ [MO_Max_Age].maxlen = 4 ;

    optdesc_ [MO_Proxy_Uri].format = OF_STRING ;
    optdesc_ [MO_Proxy_Uri].minlen = 1 ;
    optdesc_ [MO_Proxy_Uri].maxlen = 1034 ;

    optdesc_ [MO_Proxy_Scheme].format = OF_STRING ;
    optdesc_ [MO_Proxy_Scheme].minlen = 1 ;
    optdesc_ [MO_Proxy_Scheme].maxlen = 255 ;

    optdesc_ [MO_Uri_Host].format = OF_STRING ;
    optdesc_ [MO_Uri_Host].minlen = 1 ;
    optdesc_ [MO_Uri_Host].maxlen = 255 ;

    optdesc_ [MO_Uri_Path].format = OF_STRING ;
    optdesc_ [MO_Uri_Path].minlen = 0 ;
    optdesc_ [MO_Uri_Path].maxlen = 255 ;

    optdesc_ [MO_Uri_Port].format = OF_UINT ;
    optdesc_ [MO_Uri_Port].minlen = 0 ;
    optdesc_ [MO_Uri_Port].maxlen = 2 ;

    optdesc_ [MO_Uri_Query].format = OF_STRING ;
    optdesc_ [MO_Uri_Query].minlen = 0 ;
    optdesc_ [MO_Uri_Query].maxlen = 255 ;

    optdesc_ [MO_Accept].format = OF_UINT ;
    optdesc_ [MO_Accept].minlen = 0 ;
    optdesc_ [MO_Accept].maxlen = 2 ;

    optdesc_ [MO_If_None_Match].format = OF_EMPTY ;
    optdesc_ [MO_If_None_Match].minlen = 0 ;
    optdesc_ [MO_If_None_Match].maxlen = 0 ;

    optdesc_ [MO_If_Match].format = OF_OPAQUE ;
    optdesc_ [MO_If_Match].minlen = 0 ;
    optdesc_ [MO_If_Match].maxlen = 8 ;
}

/******************************************************************************
 * Accessors
 */

option::optcode_t option::optcode (void)
{
    return optcode_ ;
}

void *option::optval (int *len)
{
    *len = optlen_ ;
    return (optval_ == 0) ? staticval_ : optval_ ;
}

option::uint option::optval (void)
{
    option::uint v ;
    byte *b ;
    int i ;

    v = 0 ;
    b = (optval_ == 0) ? staticval_ : optval_ ;
    for (i = 0 ; i < optlen_ ; i++)
	v = (v << 8) & b [i] ;
    return v ;
}

/******************************************************************************
 * Mutators
 */

void option::optcode (optcode_t c)
{
    optcode_ = c;
}

void option::optval (void *val, int len)
{
    optlen_ = len ;
    COPY_VAL (val) ;
}

void option::optval (option::uint val)
{
    stbin *stbin ;

    stbin = uint_to_byte (val) ;
    CHECK_OPTLEN (optcode_, stbin->len) ;
    optlen_ = stbin->len ;
    COPY_VAL (stbin->bin) ;
}

File: pySos/sos/option.py
'''
This module contains the Option class
'''
from util.enum import Enum

class Option:
    '''
    Represents an option in a CoAP message, and it's value, if any.
    '''
    optcodes = Enum('optcodes', {'MO_NONE' : 0, 'MO_CONTENT_FORMAT' : 12,
                                 'MO_ETAG' : 4, 'MO_LOCATION_PATH' : 8,
                                 'MO_LOCATION_QUERY' : 20, 'MO_MAX_AGE' : 14,
                                 'MO_PROXY_URI' : 35, 'MO_PROXY_SCHEME' : 39,
                                 'MO_URI_HOST' : 3, 'MO_URI_PATH' : 11,
                                 'MO_URI_PORT' : 7, 'MO_URI_QUERY' : 15,
                                 'MO_ACCEPT' : 16, 'MO_IF_NONE_MATCH' : 5,
                                 'MO_IF_MATCH' : 1, 'MO_SIZE1' : 60 })

    optdesc = Enum('optdesc', ['OF_NONE', 'OF_OPAQUE', 'OF_STRING',
                               'OF_EMPTY', 'OF_UINT'])

    # Dictionary of 3-tuples
    # optdesc_[OPTCODE] = (OPT_FORMAT, OPT_MINLEN, OPT_MAXLEN)
    optdesc_ = {}

    def is_critical(self):
        '''
        Returns True if the option is critical (see CoAP draft 5.4.1)
        '''
        return self.optcode & 1

    def is_unsafe(self):
        '''
        Returns True if the option is unsafe to forward (see CoAP draft 5.4.2)
        '''
        return self.optcode & 2

    def is_nocachekey(self):
        '''
        Returns True if the option is nocachekey (see CoAP draft 5.4.2)
        '''
        return (self.optcode & 0x1E) == 0x1C

    def __init__(self, code = None, optval = None, optlen = None):
        '''
        Constructor for the option class.
        Default constructor constructs empty option.
        If you specify both optval and optlen, constructs an option with
        opaque value.
        If you specify optval but omit optlen, it is assumed that optval is an
        unsigned integer. If optval is not an unsigned integer, a 
        TypeError is raised.

        Raises : ValueError if code is invalid or optlen is specified and
                 optval is a negative integer.
                 TypeError if optlen if specified and optval is not 
                 an integer.
                 RuntimeError is the parameter combination is invalid
        '''
        # A couple of sanity checks
        if optval == None and optlen != None:
            raise RuntimeError('Invalid parameters')
        if code == None and (optval != None or optlen != None):
            raise RuntimeError('Invalid parameters')
        if optlen == None:
            if type(optval) != int:
                raise TypeError('optval is of type ' + type(optval).__name__ + 
                                ' (expected int)')
            elif optval < 0:
                raise ValueError('optval must be positive')

        if code == None or code == self.optcodes.MO_NONE:
            self.optcode = self.optcodes.MO_NONE

        self.optcode_check(code)
        self.optcode = code
        if optval == None:
            self.optlen = 0
            self.optval = None
        elif optlen == None: # Integer value
            self.optval = self.int_to_bytes(optval)
            # I'm not quite sure of this one, c++ code is kinda funky
            self.optlen = len(self.optval) 
        else: # Opaque value
            self.optlen_check(code, optlen)
            self.optval = bytearray(optval)
            self.optlen = optlen

    def __lt__(self, other):
        '''
        Comparison operator '<'. Implemented for the sole purpose of 
        sorting an option list.
        Options are sorted by option code.
        '''
        return self.optcode < other.optcode

    @staticmethod
    def optcode_check(code):
        '''
        Checks the validity of an optcode.
        Will raise a ValueError exception if it is invalid.
        '''
        if code not in Option.optdesc or code == Option.optdesc.MO_NONE:
            raise ValueError()

    @staticmethod
    def optlen_check(code, len_):
        '''
        Checks the validity of an option length. This check does not assume
        that the option code is valid, so it will raise a ValueError
        exception if the option is invalid, or if the option is valid
        but its length is not.
        '''
        if not Option.optdesc[code][1] <= len_ <= Option.optdesc[code][2]:
            raise ValueError()

    @staticmethod
    def int_to_bytes(n):
        '''
        Converts an integer into a sequence of bytes in network byte order,
        without leading null bytes.
        '''
        bytes_ = bytearray()
        while n != 0:
            b = n & 0xFF
            n = n >> 8
            bytes_.append(b)
        bytes_.reverse()
        while bytes_[0] == 0:
            del bytes_[0]
        return bytes_


# Static initialization goes here
Option.optdesc_[Option.optcodes.MO_NONE] = (OF_NONE, 0, 0)
Option.optdesc_[Option.optcodes.MO_CONTENT_FORMAT] = (OF_OPAQUE, 0, 8)
Option.optdesc_[Option.optcodes.MO_ETAG] = (OF_OPAQUE, 1, 8)
Option.optdesc_[Option.optcodes.MO_LOCATION_PATH] = (OF_STRING, 0, 255)
Option.optdesc_[Option.optcodes.MO_LOCATION_QUERY] = (OF_STRING, 0, 255)
Option.optdesc_[Option.optcodes.MO_MAX_AGE] = (OF_UINT, 0, 4)
Option.optdesc_[Option.optcodes.MO_PROXY_URI] = (OF_STRING, 1, 1034)
Option.optdesc_[Option.optcodes.MO_PROXY_SCHEME] = (OF_STRING, 1, 255)
Option.optdesc_[Option.optcodes.MO_URI_HOST] = (OF_STRING, 1, 255)
Option.optdesc_[Option.optcodes.MO_URI_PATH] = (OF_STRING, 1, 255)
Option.optdesc_[Option.optcodes.MO_URI_PORT] = (OF_UINT, 0, 2)
Option.optdesc_[Option.optcodes.MO_URI_QUERY] = (OF_STRING, 0, 255)
Option.optdesc_[Option.optcodes.MO_ACCEPT] = (OF_UINT, 0, 2)
Option.optdesc_[Option.optcodes.MO_IF_NONE_MATCH] = (OF_EMPTY, 0, 0)
Option.optdesc_[Option.optcodes.MO_IF_MATCH] = (OF_OPAQUE, 0, 8)
Option.optdesc_[Option.optcodes.MO_SIZE1] = (OF_UINT, 0, 4)
